### ISSUE TITLE
Is EF9 slower than EF8?

### ISSUE THREAD

**Issue by wallyrion** [dotnet/efcore#35053]
I observed that EF 9 is running noticeably slower and is allocating twice as much memory compared to its performance with EF 8. I'm trying to determine whether this performance degradation is an isolated exception or if there might be an issue with my benchmarking setup. Any insights would be appreciated

#### There are some benchmarks with BenchmarkDotnet

| Method          | Job      | Runtime  | UseSplitQuery | Mean     | Error     | StdDev    | Ratio | RatioSD | Allocated | Alloc Ratio |
|---------------- |--------- |--------- |-------------- |---------:|----------:|----------:|------:|--------:|----------:|------------:|
| GetCustomerById | .NET 8.0 | .NET 8.0 | False         | 1.688 ms | 0.0368 ms | 0.1038 ms |  1.00 |    0.09 |  53.82 KB |        1.00 |
| GetCustomerById | .NET 9.0 | .NET 9.0 | False         | 2.516 ms | 0.0474 ms | 0.1059 ms |  1.50 |    0.11 | 196.56 KB |        3.65 |
|                 |          |          |               |          |           |           |       |         |           |             |
| GetCustomerById | .NET 8.0 | .NET 8.0 | True          | 2.867 ms | 0.0598 ms | 0.1705 ms |  1.00 |    0.08 |  79.58 KB |        1.00 |
| GetCustomerById | .NET 9.0 | .NET 9.0 | True          | 3.114 ms | 0.0732 ms | 0.2136 ms |  1.09 |    0.10 | 161.05 KB |        2.02 |

#### Comparing the load testing metrics with k6. Performance degradation is also noticeable.

| Metric                      | .NET 8                                | .NET 9                                |
|----------------------------:|:-------------------------------------:|:-------------------------------------:|
| **Total Requests**          | 1,690,574                             | 1,287,101                             |
| **Requests per second**     | 1,741.02                              | 1,325.61                              |
| **Data Received**           | 13 GB                                 | 9.5 GB                                |
| **Data Sent**               | 220 MB                                | 164 MB                                |
| **HTTP Request Duration**   | avg=2.29s, p(90)=4.79s, p(95)=5.51s   | avg=3.32s, p(90)=6.73s, p(95)=8.06s   |
| **Request Blocking**        | avg=6.02µs, p(90)=5.01µs, p(95)=5.81µs| avg=6.65µs, p(90)=5.15µs, p(95)=6.02µs|
| **Request Receiving**       | avg=44.14µs, p(90)=61.38µs, p(95)=75.48µs| avg=42.62µs, p(90)=62.5µs, p(95)=76.01µs|
| **Request Sending**         | avg=14.77µs, p(90)=16.12µs, p(95)=32.88µs| avg=13.8µs, p(90)=15.95µs, p(95)=29.79µs|
| **Iteration Duration**      | avg=3.29s, p(90)=5.79s, p(95)=6.52s   | avg=4.32s, p(90)=7.73s, p(95)=9.06s   |
| **Virtual Users (VUs)**     | max=10,000                            | max=10,000                            |

### Query for benchmarks.

```
 var result =
            await customerQuery.Select(customer => new CustomerDto
                {
                    Id = customer.Id,
                    Name = customer.Name,
                    Orders = customer.Orders.Select(order => new OrderDto
                    {
                        OrderId = order.Id,
                        OrderDate = order.OrderDate,
                        OrderItems = order.OrderItems.Select(oi => new OrderItemDto
                        {
                            OrderItemId = oi.Id,
                            ProductName = oi.ProductName,
                            Quantity = oi.Quantity,
                            Price = oi.Price
                        }).ToList()
                    }).ToList()
                })
                .FirstOrDefaultAsync(c => c.Id == customerId);
```

#### Benchmarking ToListAsync() (retrieving 10000 items into memory)
We can observe 7x performance degradation for both speed and memory allocation between EF8 and EF9

| Method                | Job      | Runtime  | Mean    | Error    | StdDev   | Ratio | RatioSD | Allocated  | Alloc Ratio |
|---------------------- |--------- |--------- |--------:|---------:|---------:|------:|--------:|-----------:|------------:|
| GetALlCustomers       | .NET 8.0 | .NET 8.0 | 1.788 s | 0.0194 s | 0.0172 s |  1.00 |    0.01 |  272.78 MB |        1.00 |
| GetALlCustomers       | .NET 9.0 | .NET 9.0 | 7.926 s | 0.0411 s | 0.0384 s |  4.43 |    0.05 | 1784.25 MB |        6.54 |

### Complete source code for the benchmarks
[My repo](https://github.com/wallyrion/Benchmarks.EF-Dapper)

### Include provider and version information

EF Core version:
Database provider Npgsql.EntityFrameworkCore.PostgreSQL (I tried to use SqlServer provider instead of Postgres and is shows the same results)
**Version="8.0.10"
Version="9.0.0-rc.2.24474.1"** 
Target framework: (e.g. .NET 8.0 and .NET 9)
Operating system:
Windows 11 (for benchmarks) / Docker container (for load test)

**Comment by maumar** [comment]
The issue was introduced in Preview 4, specifically in this commit https://github.com/dotnet/efcore/commit/f0e88d40a4ec337c9ecf5225659c630c27a79dc7 (introduce liftable constant)

**Comment by matteomonizza** [comment]
In our project EF9 is more the 10x slower then previous

**Comment by maumar** [comment]
Example shaper in EF9:

In the benchmark, considerable amount of time is now being spent in the ShaperProcessingExpressionVisitor.CompareIdentifiers (16.5% of total time, vs 0.7% in EF8). In EF8 we used to just pass a list of ValueComparers, but in EF9 we use `List<Func<object, object, bool>>>` which we construct manually:

```
        parentIdentifierValueComparers: [LIFTABLE Constant: Func<object, object, bool>[] { Func<object, object, bool>, Func<object, object, bool> } | Resolver: _ => new Func<object, object, bool>[]
        { 
            (left, right) => left == null ? right == null : right != null && Invoke((v1, v2) => v1 == v2, (int)left, (int)right), 
            (left, right) => left == null ? right == null : right != null && Invoke((v1, v2) => v1 == v2, (int)left, (int)right) 
        }], 
        outerIdentifierValueComparers: [LIFTABLE Constant: Func<object, object, bool>[] { Func<object, object, bool>, Func<object, object, bool> } | Resolver: _ => new Func<object, object, bool>[]
        { 
            (left, right) => left == null ? right == null : right != null && Invoke((v1, v2) => v1 == v2, (int)left, (int)right), 
            (left, right) => left == null ? right == null : right != null && Invoke((v1, v2) => v1 == v2, (int)left, (int)right) 
        }], 
        selfIdentifierValueComparers: [LIFTABLE Constant: Func<object, object, bool>[] { Func<object, object, bool>, Func<object, object, bool> } | Resolver: _ => new Func<object, object, bool>[]
        { 
            (left, right) => left == null ? right == null : right != null && Invoke((v1, v2) => v1 == v2, (int)left, (int)right), 
            (left, right) => left == null ? right == null : right != null && Invoke((v1, v2) => v1 == v2, (int)left, (int)right) 
        }], 
```

Before the method was just doing a simple call to ValueComparer.Equals, but now we invoke complex linq processing involving Linq.Expressions.Interpreter

**Comment by matteomonizza** [comment]
In EF9, a query in SQLite that perform in EF8 in one second, now takes 12 seconds.. we are investigating..

**Comment by maumar** [comment]
Benchmarks after removing invoke from the ValueComparer.ObjectEqualsExpression:

ef 8

|                    Method | Async |     Mean |   Error |  StdDev |  Op/s |      Gen0 |      Gen1 | Allocated |
|-------------------------- |------ |---------:|--------:|--------:|------:|----------:|----------:|----------:|
| PredicateMultipleIncludes | False | 147.2 ms | 2.63 ms | 2.46 ms | 6.793 | 4000.0000 | 3000.0000 |  26.24 MB |
| PredicateMultipleIncludes |  True | 159.1 ms | 3.00 ms | 2.95 ms | 6.287 | 5500.0000 | 3000.0000 |  34.47 MB |

ef 9 without this change

|                    Method | Async |     Mean |   Error |  StdDev |  Op/s |       Gen0 |      Gen1 | Allocated |
|-------------------------- |------ |---------:|--------:|--------:|------:|-----------:|----------:|----------:|
| PredicateMultipleIncludes | False | 322.6 ms | 0.97 ms | 0.86 ms | 3.099 | 13000.0000 | 6000.0000 |  79.48 MB |
| PredicateMultipleIncludes |  True | 344.9 ms | 6.79 ms | 6.67 ms | 2.899 | 14000.0000 | 7000.0000 |  87.72 MB |

ef 9 with this change

|                    Method | Async |     Mean |   Error |  StdDev |  Op/s |       Gen0 |      Gen1 | Allocated |
|-------------------------- |------ |---------:|--------:|--------:|------:|-----------:|----------:|----------:|
| PredicateMultipleIncludes | False | 242.8 ms | 2.39 ms | 2.12 ms | 4.119 |  8000.0000 | 5000.0000 |  51.69 MB |
| PredicateMultipleIncludes |  True | 263.4 ms | 2.21 ms | 2.06 ms | 3.797 | 10000.0000 | 9000.0000 |  59.93 MB |

Significant improvement, but still doesn't fully address the problem. Looking for more...

**Comment by zakeryooo** [comment]
I've just hit what I assume is this issue too. Updated to EF9 and compared to EF Core 8 when dealing with medium+ sized data sets it's very slow. 

Loading a full table with 2 joins into memory takes 1.5s in EF8, it takes 9s for the same data in EF9. Saving changes to a small subset of that data in EF8 takes 0.5s, it takes 20s in EF9, same data.

Really surprised this issue hasn't had a lot more comments, it seems like a huge regression?

**Comment by matteomonizza** [comment]
EF9 is killing my app.. From 1s to 10s is to much...  I use DataReader and DataTable loading via DataReader, think the problem is here...

**Comment by maumar** [comment]
@zakeryooo are your queries using collection navigation? If not, could you share the query, model that is involved and the approximate size of the data in your database? Also, I'm interested in the update scenario - can you provide some more info on what you are doing (small repro code would be ideal). If you experience catastrophic perf regression on EF9 it makes sense to stay on EF8 for now. This is a high priority issue for us and we hope to have a significant improvement in the first patch release. But we really want to make sure we understand all the scenarios that are impacted - the more info users provide the better

**Comment by maumar** [comment]
@matteomonizza can you provide some more info on your scenario? What query shapes are involved and what is the size of data that you are working with? Standalone repro would be the best, but any info would be helpful - we want to make sure that root cause you encountered in the same and the one reported originally.

**Comment by zakeryooo** [comment]
> [@zakeryooo](https://github.com/zakeryooo) are your queries using collection navigation? If not, could you share the query, model that is involved and the approximate size of the data in your database? Also, I'm interested in the update scenario - can you provide some more info on what you are doing (small repro code would be ideal). If you experience catastrophic perf regression on EF9 it makes sense to stay on EF8 for now. This is a high priority issue for us and we hope to have a significant improvement in the first patch release. But we really want to make sure we understand all the scenarios that are impacted - the more info users provide the better

Yeah it's using collection navigation, I'm away at the moment so unfortunately have limited ability to provide particularly useful info but the general gist of the code is:

            var products = _ctx.Products
                .AsSplitQuery()
                .Include(p => p.ShopEntries)
                    .ThenInclude(se => se.ActivePrices)
                .Include(p => p.ShopEntries)
                    .ThenInclude(se => se.Shop)
                    .ToList();
            foreach (var product in products)
            {
                decimal popularity = 0;
                foreach (var shopEntry in product.ShopEntries)
                {
                    if (shopEntry.ActivePrices.Count == 0 || shopEntry.Shop.State == ShopState.Inactive) continue;
                    if (shopEntry.ActivePrices.All(ap => ap.CurrentPrice == null)) continue;

                    // Basic logic here to update the popularity value, each run usually only changes the value for < 100 records
                }
                product.ComputedPopularity = (int)popularity;
            }
            _ctx.SaveChanges();

The actual SQL queries only take ~100ms total on the DB side, 22k products, 100k shop entries, 42k active prices, 22 shops. 

On the exact same data / popularity changes:

Warm cached on EF8 takes:
  ToList Products: 2.5s - Loop Products: 20ms - SaveChanges: 0.5s

Warm cached on EF9 takes: 
  ToList Products: 8s - Loop Products: 20ms - SaveChanges: 20s. 

I'm back next week and if you still need it by then can put together a small repro.

**Comment by matteomonizza** [comment]
> [@matteomonizza](https://github.com/matteomonizza) can you provide some more info on your scenario? What query shapes are involved and what is the size of data that you are working with? Standalone repro would be the best, but any info would be helpful - we want to make sure that root cause you encountered in the same and the one reported originally.

Yesterday we took time to face the problem.. The app is loading data from a SQLite database. The app is an iOS and Android app (not MAUI, we use Net-iOS and Net-Android). The Microsoft SQLite provider we use is the Microsoft.Data.Sqlite, that does not contain the DataAdapter implementation. We create out implementation and now the LoadData of the data table take s only 1s against 10s with fill from data reader (with Net 8 there was not this gap). Then, we notice that using the app in debug mode, both Android and iOS version both simulator and device, is extremely slow. For example, the same code run in 2s in release mode, in 30s in debug mode, only on .Net 9. Tried this on different machines.

**Comment by c5racing** [comment]
My situation may or may not be a bit different.  We use PostgreSQL and are in the process of migrating from .NET8 to .NET9 and in doing so, changing from the Legacy POCO mapping to the newer `.ToJson()` Mapping.

I've attached a sample project (thrown together very quickly) using the identical data.  Neither one actually modifies data.  The legacy POCO mapping takes 0ms to call SaveChanges().  The newer `.ToJson()` takes over 2 seconds.

[https://drive.google.com/file/d/1A_VoN9XgrogMEeevV7mtLh5c2s-Ds2rt/view?usp=sharing](https://drive.google.com/file/d/1A_VoN9XgrogMEeevV7mtLh5c2s-Ds2rt/view?usp=sharing)

**Comment by roji** [comment]
@c5racing thanks, this is something that's definitely worth checking once we release some planned fixed to help the general performance problem (for 9.0.1). One thing to confirm though - are you testing both scenarios on EF 9 (i.e. both ToJson and legacy POCO mapping)?

**Comment by c5racing** [comment]
> [@c5racing](https://github.com/c5racing) thanks, this is something that's definitely worth checking once we release some planned fixed to help the general performance problem (for 9.0.1). One thing to confirm though - are you testing both scenarios on EF 9 (i.e. both ToJson and legacy POCO mapping)?

Yes, testing both scenarios on EF9.  Calling `var hasChanges = newDbContext.ChangeTracker.HasChanges();` returns false but `SaveChanges()` still takes 2-3 second

**Comment by roji** [comment]
Ah, if this is about a regression in SaveChanges() performance, then can you please open a new issue for it? This issue is about regressions specifically in reading back results from queries, which is quite a different thing.

**Comment by f0ppa21** [comment]
We are also experiencing that EF 9 are much slower than EF 8 for queries that result in large / complex datamodels. The generated Sql-query (split query) performs similar as for EF 8, its the subsequent call to **ToListAsync** that seems to execute much slower (x2 slower for only 2700 returned records and gets worse with more returned records).

**Comment by maumar** [comment]
@f0ppa21 we have made several improvements that will ship in 9.0.1 patch. The perf is not exactly back to EF8 levels but should be close. In the meantime, you can try our daily builds to see how the upcoming improvements affect your particular scenario. If you still see significant perf degradation using daily builds, we would appreciate the repro so we can investigate and target your specific case.

**Comment by FredSwadlingPelt8** [comment]
Maybe this may be of help to anyone else struggling with the ef core update from v8 to v9; Not sure if it relates at all to the above, but we found that after the upgrade EF Core started defaulting to translating parametrized collections to json which was killing performance on a lot of our queries. Easy fix by adding TranslateParameterizedCollectionsToConstants() to the options builder to return it to its previous behavior.

https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.infrastructure.relationaldbcontextoptionsbuilder-2.translateparameterizedcollectionstoconstants?view=efcore-9.0

**Comment by roji** [comment]
@FredSwadlingPelt8 the change to translating parameterized collections to JSON was introduced in 8, not 9; the issue tracking this is #32394.

**Comment by FredSwadlingPelt8** [comment]
> [@FredSwadlingPelt8](https://github.com/FredSwadlingPelt8) the change to translating parameterized collections to JSON was introduced in 8, not 9; the issue tracking this is [#32394](https://github.com/dotnet/efcore/issues/32394).

I did check, and the parameterized query issue only appeared in our queries when we updated from ef core 8 to ef core 9. I have no idea why this would be.

**Comment by maumar** [comment]
Added some more improvements around SaveChanges scenarios with primitive collections. This issue should now be fixed - if you still see issues after updating to EF 9.0.2 please file a new issue with details about the scenario.
